### PR TITLE
DDT-375+378: Implement ShipmentLocation and ShipmentCutOffTimes deep object

### PR DESCRIPTION
**Pull request by mp-assco**

### CODE REVIEW

File: src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentCutOffTimeTO.java [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
+import org.springframework.data.annotation.Id;
+import org.springframework.data.relational.core.mapping.Column;
+import org.springframework.data.relational.core.mapping.Table;
 
 import javax.validation.constraints.NotNull;
 import java.time.OffsetDateTime;
+import java.util.UUID;
 
 @Data
+@Table("shipment_cutoff_time")
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentCutOffTimeTO.java:16]: Separate this from the TO class, create a model for the table in event-core.

File: src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java [hunk 2]
```diff
@@ -8,19 +8,27 @@
...
 
 @Data
 public class ShipmentLocationTO {
+  @NotNull(message = "ShipmentID is required.")
+  private UUID shipmentID;
+
+  @NotNull(message = "BookingID is required.")
+  private UUID bookingID;
+
+  private String LocationID;
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java:21]: Not sure if `LocationID` is required here, I don't see it in the `shipmentLocation` object on swagger.

File: src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java [hunk 3]
```diff
@@ -8,19 +8,27 @@
 import javax.validation.constraints.NotNull;
 import javax.validation.constraints.Size;
 import java.time.OffsetDateTime;
+import java.util.UUID;
 
 @Data
 public class ShipmentLocationTO {
+  @NotNull(message = "ShipmentID is required.")
+  private UUID shipmentID;
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java:16]: Not sure if `shipmentID` is required here, I don't see it in the `shipmentLocation` object on swagger.

File: src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java [hunk 4]
```diff
@@ -8,19 +8,27 @@
...
+
+  @NotNull(message = "BookingID is required.")
+  private UUID bookingID;
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentLocationTO.java:19]: Not sure if `bookingID` is required here, I don't see it in the `shipmentLocation` object on swagger.

File: src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package org.dcsa.bkg.repositories;
+
+import org.dcsa.core.events.model.Shipment;
+import org.dcsa.core.repository.ExtendedRepository;
+import reactor.core.publisher.Mono;
+
+public interface BookingConfirmationRepository
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java:7]: `BookingConfirmationRepository` is the same as `ShipmentRepository` which exists in event core, maybe use that instead.

File: src/main/java/org/dcsa/bkg/repositories/CarrierClauseRepository.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package org.dcsa.bkg.repositories;
+
+import org.dcsa.bkg.model.transferobjects.CarrierClauseTO;
+import org.dcsa.core.repository.ExtendedRepository;
+
+import java.util.UUID;
+
+public interface CarrierClauseRepository extends ExtendedRepository<CarrierClauseTO, UUID> {
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/repositories/CarrierClauseRepository.java:8]: Move to event-core.

File: src/main/java/org/dcsa/bkg/repositories/ShipmentCutOffTimeRepository.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package org.dcsa.bkg.repositories;
+
+import org.dcsa.bkg.model.transferobjects.ShipmentCutOffTimeTO;
+import org.dcsa.core.repository.ExtendedRepository;
+import reactor.core.publisher.Flux;
+
+import java.util.UUID;
+
+public interface ShipmentCutOffTimeRepository
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/repositories/ShipmentCutOffTimeRepository.java:9]: Move to event-core, and use an entity `ShipmentCutOffTime` instead of the TO.

**Review by douradofrazer — CHANGES_REQUESTED** [review]

File: src/main/java/org/dcsa/bkg/repositories/CarrierClauseRepository.java (hunk 6 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/bkg/repositories/CarrierClauseRepository.java:8]: https://github.com/dcsaorg/DCSA-Event-Core/pull/142

File: src/main/java/org/dcsa/bkg/model/transferobjects/BookingConfirmationTO.java [hunk 8]
```diff
@@ -31,5 +34,5 @@
 
   private List<ChargeTO> charges;
 
-  private List<CarrierClauseTO> carrierClauses;
+  private List<CarrierClause> carrierClauses;
```
Comment (pimg) [inline src/main/java/org/dcsa/bkg/model/transferobjects/BookingConfirmationTO.java:37]: we are now mixing TO objects and data objects in the TO response. Would be better if a TO object is made from other TO objects and not include data objects

Comment (mp-assco) [reply src/main/java/org/dcsa/bkg/model/transferobjects/BookingConfirmationTO.java:37]: Good point, I'll do that.

File: src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java (hunk 5 above)
Comment (pimg) [inline src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java:7]: is this file still needed? Seems like it is not used and instead the ShippingRepository from Event Core is used.

**Review by pimg — CHANGES_REQUESTED** [review]
one comment regarding the confirmedBookingRepository

File: src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java (hunk 5 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/bkg/repositories/BookingConfirmationRepository.java:7]: Ah, you're right. I've deleted it.

**Review by pimg — APPROVED** [review]

File: src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java [hunk 9]
```diff
@@ -950,6 +981,85 @@ void testGETBookingWithLocationDeepAndCommoditiesAndValAddedServAndRefsAndReqEqs
   }
 
   @Nested
+  @DisplayName("")
```
Comment (douradofrazer) [inline src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java:984]: The display name is missing.

**Review by douradofrazer — CHANGES_REQUESTED** [review]

File: src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentCutOffTimeTO.java [hunk 10]
```diff
@@ -1,15 +1,22 @@
...
+import org.springframework.data.annotation.Id;
 
 import javax.validation.constraints.NotNull;
 import java.time.OffsetDateTime;
+import java.util.UUID;
 
 @Data
 public class ShipmentCutOffTimeTO {
 
+  @Id
```
[the remark this answers is not in the document]
Comment (mp-assco) [reply src/main/java/org/dcsa/bkg/model/transferobjects/ShipmentCutOffTimeTO.java]: @douradofrazer A mistake. It's been removed.

File: src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java (hunk 9 above)
Comment (mp-assco) [reply src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java]: @douradofrazer Updated to reflect test.

**Review by pimg — DISMISSED** [review]

**Review by douradofrazer — APPROVED** [review]
